index.d.ts

```diff
@@ -0,0 +1,888 @@
+  rowIndex: number;
+  /**
+   * @description
+   * full A1 formatted range of this row
+   * - includes the worksheet name, ex: "sheet1!A5:D5"
+   */
+  a1Range: string;
+}
+
+/**
+ * Property keys are determined by the header row of the sheet
+ * - each row will have a property getter/setter available for each cell corresponding to the column header
+ */
+interface GoogleSpreadsheetRow extends RowProperties {
+  /**
+   * @description
+   * save any updates made to cell values in this row
+   */
+  save(): Promise<void>;
+  /**
+   * @description
+   * delete this row
+   */
+  delete(): Promise<void>;
+}
+
+/* ---- GOOGLE SPREADSHEET WORKSHEET ---- */
+
+interface WorksheetBasicProperties {
+  /**
+   * ---- BASIC PROPERTIES ----
+   * separates basic (editable) properties as they are used as inputs to various methods
+   * non-basic properties should be added to the extending interface below
+   */
+
+  /**
+   * @description
+   * name of the worksheet tab
+   */
+  title: string;
+  /**
+   * @description
+   * tab index in the worksheet doc (based on rightToLeft property)
+   */
+  index: number;
+  /**
+   * @description
+   * additional properties of the worksheet if this sheet is a grid
+   */
+  gridProperties: WorksheetGridProperties;
+  /**
+   * @description
+   * true if the worksheet is hidden in the UI, false if it's visible
+   */
+  hidden: boolean;
+  /**
+   * @description
+   * the color of the worksheet tab
+   */
+  tabColor: Color;
+  /**
+   * @description
+   * true if the worksheet is an RTL sheet instead of an LTR sheet
+   */
+  rightToLeft: boolean;
+}
+
+interface GoogleSpreadsheetWorksheet extends WorksheetBasicProperties {
+  /* ---- NON-BASIC PROPERTIES ADDED HERE ---- */
+
+  /**
+   * @description
+   * set during creation, not editable
+   */
+  readonly sheetId: string;
+  /**
+   * @description
+   * set during creation, not editable
+   */
+  readonly sheetType: WorksheetType;
+  /**
+   * @description
+   * number of rows in the worksheet
+   */
+  rowCount: number;
+  /**
+   * @description
+   * number of columns in the worksheet
+   */
+  columnCount: number;
+  /**
+   * @description
+   * alias for the worksheet title
+   */
+  a1SheetName: string;
+  /**
+   * @description
+   * A1 column letter of the last column in the worksheet
+   */
+  lastColumnLetter: string;
+  /**
+   * @description
+   * stats about the cells in the worksheet
+   */
+  cellStats: CellStats;
+
+  /* ---- SYNCHRONOUS METHODS ---- */
+
+  /**
+   * @description
+   * retrieve a cell from the cache based on numeric row and column indices
+   */
+
+  getCell(rowIndex: number, columnIndex: number): GoogleSpreadsheetCell;
+  /**
+   * @description
+   * retrieve a cell from the cache based on A1 address
+   */
+  getCellByA1(a1Address: string): GoogleSpreadsheetCell;
+  /**
+   * @description
+   * fetch all rows in the worksheet
+   *
+   * @param options
+   */
+  getRows(options: PaginationOptions): [GoogleSpreadsheetRow];
```

Comment:
A few notes:

- The syntax `[GoogleSpreadsheetRow]` means a tuple with a single element of type `GoogleSpreadheetRow` and not an array of `GoogleSpreadsheetRow` like `GoolgeSpreadsheetRow[]`.
- The return value is expected to be a Promise according to the [doc](https://theoephraim.github.io/node-google-spreadsheet/#/classes/google-spreadsheet-worksheet?id=fn-getrows).
- The `options` are optional.

Reply:
thanks man. im a bit confused. i put up this PR but then noticed a similar PR went up a day after mine. @theoephraim hasnt responded to this one (but has been commenting on the other).

Reply:
thanks man. im a bit confused. i put up this PR but then noticed a similar PR went up a day after mine. @theoephraim hasnt responded to this one (but has been commenting on the other).

wondering if i should fix these or abandon it. kind of a bummer man. i spent all night researching how TS works and writing all 900 lines of type defs by hand

Reply:
for my own information how would you define something like this as an array instead of a tuple?

```
/**
   * @description
   * loads the cells in the worksheet
   * - strings are treated as an A1 range
   * - objects are treated as a WorksheetGridRange
   *
   * @param filters single or array of filters
   */
  loadCells(
    filters: string | WorksheetGridRange | [string | WorksheetGridRange],
  ): Promise<void>;
```

like this (is there a more concise way)?

```ts
/**
   * @description
   * loads the cells in the worksheet
   * - strings are treated as an A1 range
   * - objects are treated as a WorksheetGridRange
   *
   * @param filters single or array of filters
   */
  loadCells(
    filters: string | WorksheetGridRange | string[] | WorksheetGridRange[],
  ): Promise<void>;
```

Reply:
it was simple enough so i  pushed up a fix to those tuples that should have been arrays (there were a few of them spread throughout). and fixed `getRows` to have optional options and return `Promise<GoogleSpreadsheetRow[]>`

Reply:
don't fret :) The work is much appreciated, and I'm sure you learned a bunch in the process! From a cursory look, the other one also used `enum` and `class` which I thought was maybe the better way to do things? But I'm not a big TS user, so I don't really know. My plan was to review them both a bit and hopefully come out with a final version that combines them.

Regardless, it's much appreciated, and just bad timing that someone else also probably stayed up all night doing the same thing!

Reply:
ah okay no worries man. it was definitely a good learning experience for the future. 

maybe @EndBug or @vogre can comment on which is preferred. once again i have absolutely no idea what i was doing lol.

my focus was on making it work as expected (to have type support when using the JS lib) moreso than writing them as traditional TS (for actual TS usage). 

heres what i learned when ending up with this format:

- declare module meant that everything inside it was exported (so intellisense was autocompleting things that shouldnt be visible)
- it wont let you use `enum` without declaring / exporting it (problem from above). but the `type` with strings did work to autocomplete those values as soon as you opened a `"` quote
- like the module problem using `class` seems to automatically export it. so i only defined classes for the 2 exports the docs suggest using and left interfaces for the others. that way you still get typings but dont expose exports that shouldnt be used

Reply:
i see value in the way @EndBug wrote it if you intend to transition to TS in the future because it lays out all the groundwork including the private fields / methods. but, at least in my own very limited experience, led to unexpected behavior as typings for a native JS lib. 

i had to use a lot of trial and error to get it working as expected for typings. so i have no doubt some of it is questionable (to say the least lol) in the traditional TS sense.

Reply:
Hey, I've seen your PR at the time, but I decided to publish a new one anyway since I had pretty much ready 😅 
I've seen the problem with using `enum` and I have to admit that it is not very practical to always import the enum before defining it: I didn't think bout that when I was writing it, to be honest...
The problem is also that these enums are not really exported, they're only exported as definition: that means that they can't even be used because it would result in an "undefined property" error when running the compiled version:
```js
var a = require("google-spreadsheet").Dimension.COLUMNS;
// TypeError: Cannot read property 'COLUMNS' of undefined
```

@theoephraim What should I do, should I convert all of them to string literals?

Reply:
ya thats what i was referring to. it seems in general there are several conflicts with writing proper TS when used for typings in a native JS lib. 

my thoughts are:
- **short term:** merge this PR because it will immediately give all users consistent and non-conflicting behavior for v3 (google API v4) typings when used with the existing JS lib. 
- **long term:** if @theoephraim is keen on migrating to TS in the future then use @EndBug branch as a basis for that rewrite

im actually pretty interested in learning more about TS now that i got a taste for it. seems there are numerous benefits for long term stability of the lib. the google sdk and docs are horrendous and i wouldnt even consider working with the sheets API if it werent for this lib. id love to contribute to ensuring its stability and growth.

of course this will take a lot of time. but if @EndBug and @theoephraim are up for it i would be willing to collaborate on converting the lib entirely to TS from the source to the tests and docs.

Reply:
If it was my decision I would definitely switch to TS, since once you do it then you don't have to worry about updating the typings and it's a better overall experience for me. 
There's some work to be done but it shouldn't be too much: once you set up the structure you don't have to explicitly type everything right away since everything would still work the same even without types. 
It is also true that not all functions need explicit typing, a lot of them will be automatically generated by TS. In my experience, it can be easily done without a lot of problems.

Reply:
thats a good point man. from what i read it is technically a superset of JS. so it should be possible to slowly migrate it (if we can ignore the yellings of the linter lol). but i think it would require configuring a transpiler and some other overhead right?

correct me if im wrong, but another benefit would be auto-generated docs. if we move to TS and add JSDoc annotations then the docs would update as the code changes instead of having to synchronize them separately. 

thinking objectively can you identify any problems or requirements to migrating it? any other overhead that @theoephraim should be aware of in considering it? i dont want to speak for him but it sounds like he is about as green as me with TS from the comments ive seen. im sure hed need to know all the details to make a decision one way or the other.

Reply:
In TypeScript you don't have to explicitly type anything, and so it shows no errors when the `any` type is used. The errors come if there's some kind of code that work only because of type coercion or stuff like that, but you can solve it right away by putting a type.

You don't need any external compiler: installing the `typescript` package adds the TypeScript Compiler with the `tsc` CLI, which can be used for the build. I've talked a bit about that [here](https://github.com/theoephraim/node-google-spreadsheet/pull/289#issuecomment-584793823)

Using JSDoc and TS adds a sort of inline docs, but I don't like the idea of removing the docs entirely: in my opinion, it's nice to have a nicely formatted docs page that shows you everything you need to know in a clean way. Without it, you would need to cycle through everything IntelliSense sees and it would be very unpractical.

Reply:
oh thats really neat.

> but I don't like the idea of removing the docs entirely

i agree. what i meant was that the properties / methods can be written into the docs site files instead of having to manually synchronize them. for example the properties and methods